### PR TITLE
Parser: Report a Warning if implicit integer conversion truncates value or sign conversion.

### PR DESCRIPTION
**Pull request by Organ1sm**
When I wanted to learn how to write a C23 C language compiler, this project was extremely helpful to me, no less than having a rich big meal right after finishing a workout. I've been going through this project almost commit by commit, although I haven't finished looking through it all yet. This project not only taught me how to write a C compiler for the C23 standard, but also allowed me to learn a new language, Ziglang, which made me feel very happy. So I want to see if I can do something for this project. I looked at the issue section, and because I don't have experience contributing to open-source projects, I can only start with issues labeled as good first issue.

so I attempted to solve this issue.  #701 

Value.intCast` currently  return a value that indicates whether truncation happens and then issue a diagnostic.

### CODE REVIEW

File: src/aro/Parser.zig [hunk 1]
```diff
@@ -5574,7 +5574,11 @@ pub const Result = struct {
                 try res.implicitCast(p, .complex_float_to_complex_int);
             }
         } else if (!res.ty.eql(int_ty, p.comp, true)) {
-            try res.val.intCast(int_ty, p.comp);
+            const old_val = res.val;
+            const value_change_kind = try res.val.intCast(int_ty, p.comp);
+            if (value_change_kind == .value_changed and !(res.ty.makeIntegerUnsigned().specifier == int_ty.makeIntegerUnsigned().specifier))
```
Comment (ehaas) [inline src/aro/Parser.zig:5579]: Is there a specific scenario the `!(res.ty.makeIntegerUnsigned().specifier == int_ty.makeIntegerUnsigned().specifier)` part guards against (probably better expressed using `!=` instead)?

**Comment by ehaas** [comment]
The windows `binary expressions.c` test is failing because `long` is only 4 bytes on windows, so `0xFFFFFFFFFF` gets promoted to `long long`. Changing line 48 to `int baz = 0xFFFFFFFFFFLL + 1u;` and updating the error text to `long long` will make it have the same behavior on all platforms.

**Comment by ehaas** [comment]
For the 2 new warnings in `enum overflow.c` it looks like `C` and `H` really do have type `int` on windows (confirmed on godbolt https://godbolt.org/z/x4PdsKEcs) - so for those two it looks like we should be comparing against `-1` in the windows static asserts

**Comment by ehaas** [comment]
BTW thank you for the kind words and contribution!

**Comment by Organ1sm** [comment]
According to your advice, I revised the code and pushed it. Thank you for your code review. :)

**Comment by Organ1sm** [comment]
Hello, it seems this PR hasn't shown any signs of being merged. I'd like to know if there are any parts of the code that aren't well written so I can continue to improve them.

**Comment by ehaas** [comment]
I had a question about the `res.ty.makeIntegerUnsigned().specifier != int_ty.makeIntegerUnsigned().specifier` condition - it's not clear to me why that's needed. Can you give an example that does the wrong thing if that condition is not included?

**Comment by Organ1sm** [comment]
For example:   ` unsigned int u = -1;`
The left-hand side and right-hand side of this statement only differ in terms of being unsigned and signed. At the semantic level, there is an implicit conversion. So initially, my approach to solving this was not to add this extra condition, and I also wanted to add a warning for this implicit conversion. However, I later found that both [Clang](https://godbolt.org/z/s78rYPEhe) and [GCC](https://godbolt.org/z/9Wvhdvvcf) do not issue warnings for this statement.

**Comment by ehaas** [comment]
I agree that adding a `sign-conversion` diagnostic option kind is the way to go. The bigint `bitCountTwosComp` function may be able to help with figuring out if the value changed due to truncation vs sign change.

File: src/aro/Value.zig [hunk 2]
```diff
@@ -230,23 +230,51 @@ pub fn intToFloat(v: *Value, dest_ty: Type, comp: *Compilation) !void {
...
+
+    const dest_bits: usize = @intCast(dest_ty.bitSizeof(comp).?);
+    const dest_signed = dest_ty.signedness(comp) == .signed;
+
     var space: BigIntSpace = undefined;
     const big = v.toBigInt(&space, comp);
+    const value_bits = big.bitCountTwosComp();
+
+    // if big is negative or zero, then is signed.
+    const src_signed = (!big.positive or big.limbs[0] == 0);
```
Comment (ehaas) [inline src/aro/Value.zig:255]:
```suggestion
    const src_signed = (!big.positive or big.eqlZero());
```

**Comment by ehaas** [comment]
Looks good to me!

File: src/aro/Value.zig [hunk 3]
```diff
@@ -230,23 +230,51 @@ pub fn intToFloat(v: *Value, dest_ty: Type, comp: *Compilation) !void {
...
+
+    const dest_bits: usize = @intCast(dest_ty.bitSizeof(comp).?);
+    const dest_signed = dest_ty.signedness(comp) == .signed;
+
     var space: BigIntSpace = undefined;
     const big = v.toBigInt(&space, comp);
+    const value_bits = big.bitCountTwosComp();
+
+    // if big is negative or zero, then is signed.
+    const src_signed = (!big.positive or big.eqlZero());
```
Comment (Vexu) [inline src/aro/Value.zig:255]: Why is zero considered signed? Clang and gcc don't do so.

**Review by Vexu — COMMENTED** [review]
Thanks for the enhancement and I'm glad the project has been helpful to you!

File: src/aro/Value.zig (hunk 3 above)
Comment (Organ1sm) [reply src/aro/Value.zig]: yes，I tested the behavior of clang and gcc. It was my oversight, so I wrote a new commit to correct this.

**Pull request merged**